rclcpp/src/rclcpp/clock.cpp

```diff
@@ -76,6 +90,77 @@ Clock::now()
   return now;
 }
 
+bool
+Clock::sleep_until(Time until)
+{
+  const auto this_clock_type = get_clock_type();
+  if (until.get_clock_type() != this_clock_type) {
+    RCUTILS_LOG_ERROR("sleep_until Time clock type does not match this clock's type.");
+    return false;
+  }
+  bool time_source_changed = false;
+
+  if (this_clock_type == RCL_STEADY_TIME) {
+    auto steady_time = std::chrono::steady_clock::time_point(
+      std::chrono::nanoseconds(until.nanoseconds()));
+
+    // loop over spurious wakeups but notice shutdown
+    std::unique_lock lock(impl_->clock_mutex_);
+    while (now() < until && rclcpp::ok()) {
+      impl_->cv_.wait_until(lock, steady_time);
+    }
+  } else if (this_clock_type == RCL_SYSTEM_TIME) {
+    auto system_time = std::chrono::time_point<
+      std::chrono::system_clock, std::chrono::nanoseconds>(
+      std::chrono::nanoseconds(until.nanoseconds()));
+
+    // loop over spurious wakeups but notice shutdown
+    std::unique_lock lock(impl_->clock_mutex_);
+    while (now() < until && rclcpp::ok()) {
+      impl_->cv_.wait_until(lock, system_time);
+    }
+  } else if (this_clock_type == RCL_ROS_TIME) {
+    // Install jump handler for any amount of time change,
+    // to check if time has been reached on each new sample
+    rcl_jump_threshold_t threshold;
+    threshold.on_clock_change = true;
+    threshold.min_backward.nanoseconds = 0;
+    threshold.min_forward.nanoseconds = 0;
+    auto clock_handler = create_jump_callback(
```

Comment:
No, we need this jump handler to detect when `ros_time_is_active` changes. The `on_clock_change = true` makes it so that the callback is triggered when ros_time changes in either direction. So, if it was really system time, but ros time becomes active, we want to quit sleeping because system time is no longer relevant.